ideal_indirection.md

```diff
@@ -87,80 +98,84 @@ Photo Cred: http://duartes.org/gustavo/blog/post/anatomy-of-a-program-in-memory/
+For this assignment all the physical memory allocations will be abstracted by `kernel.c`.
 
 This file will maintain a global array of pages that you will use to model all of physical memory.
 That is to say that all virtual addresses get translated to an address in:
 
 `char physical_memory[PHYSICAL_MEMORY_SIZE] __attribute__((aligned(PAGE_SIZE)));`
 
+This array of physical memory will be divided into frames of size `PAGE_SIZE`.
+The kernel comes with functionality to
+- request and free up frames
+- swap frames to and from disk
+
+You will be using these functions to obtain physical memory frames for the page tables and user data, as well as reading pages that have been swapped out to disk.
+
 The caveat to this lab is that it is all done in user space.
-That means you are technically mapping one virtual address to another virtual address that represents a physical address.
-However, all the concepts involved remain the same in a real operating system's memory management software.
+That means you are technically mapping a virtual address in the simulated user space to a virtual address in your user space in the actual machine. However, all the concepts involved remain the same in a real operating system's memory management software. Thus, the address translation is as follows:
 
-We use a global `char` array for our physical memory as it so happens that global variables such as these are stored
-in some of the lowest addresses in memory. Because of this, the array, despite existing in a 64 bit environment, only
-needs the 32 lower bits of a 64 bit address to address it. This is great because it allows us to use 32 addresses to
-refer to a physical memory location, despite being on a 64 bit system. The downside is that we will need to convert
-any 32 bit references to the data into a 64 bit pointer before we actually try to access the physical memory at that
-address.
+Simulated virtual address $$\rightarrow$$ Simulated physical address $$\rightarrow$$ Virtual address $$\rightarrow$$ Physical address
 
-In this lab, we will be dealing with `addr32` type variables to keep track of both virtual and "physical" addresses.
-The type `addr32` is just an alias for `uint32_t`, which is just a type that can store a 32 bit value. However,
-since our actual VMs are 64 bit, we will need to convert this `addr32` into a system pointer if we ever want to interact
-with real physical memory directly (ie, by dereferencing a pointer). We have provided a few helper functions to 
-help with this translation: 
+Note that the final transation to actual physical address will be done by the system.
 
+We use a global `char` array for our simulated physical memory as it so happens that global variables such as these are stored in some of the lowest addresses in virtual memory.
+Because of this, the array, despite existing in a 64 bit environment, only needs the 32 lower bits of a 64 bit address to address it.
```

Comment:
```suggestion
Because of this, the array, despite existing in a 64-bit environment, only needs the 32 lower bits of a 64-bit address to address it.
```